### PR TITLE
HSEARCH-2510 ArrayBridge doesn't support arrays of primitive types

### PR DESCRIPTION
**Pull request by yrodiere**
This PR is based on #1273 (HSEARCH-2535), which will have to be merged first.

Relevant ticket: https://hibernate.atlassian.net/browse/HSEARCH-2510

Not sure it's the right time to merge this, but I had to investigate and found a solution while working on HSEARCH-2535, so I though we might as well discuss it.

~~The main drawback of the solution is performance: it appears `Array.*` static methods are notoriously slow, as [discussed there](http://stackoverflow.com/questions/30306160/performance-of-java-lang-reflect-array) and [reported there](https://bugs.openjdk.java.net/browse/JDK-8051447]. If it's an issue, I can set up our own utils in Java: it appears that would be faster. Not sure that's worth it, though: yes, these methods are slower than using the built-in equivalents (`[]` operator and `array.length`) but I suppose the latter are pretty fast to begin with.~~ => Addressed by adding utils to `CollectionHelper`.

### CODE REVIEW

File: engine/src/main/java/org/hibernate/search/engine/spi/DocumentBuilderIndexedEntity.java [hunk 1]
```diff
@@ -677,7 +685,9 @@ else if ( member.isCollection() && currentFieldValue instanceof Map ) {
 							}
 							else if ( member.isArray() ) {
 								multivalued = true;
-								for ( Object element : (Object[]) currentFieldValue ) {
+								// Use Array.* to also support arrays of primitive values
+								for ( int i = 0, length = Array.getLength( currentFieldValue ); i < length; ++i ) {
```
Comment (gsmet) [inline engine/src/main/java/org/hibernate/search/engine/spi/DocumentBuilderIndexedEntity.java]: I think we usually use `i++`

File: engine/src/main/java/org/hibernate/search/engine/spi/DocumentBuilderIndexedEntity.java [hunk 2]
```diff
@@ -635,6 +640,9 @@ else if ( currentFieldValue instanceof Map ) {
 						objectInitializer.initializeMap( (Map) currentFieldValue );
 					}
 				}
+				else if ( member.isArray() && currentFieldValue instanceof Object[] ) {
```
Comment (gsmet) [inline engine/src/main/java/org/hibernate/search/engine/spi/DocumentBuilderIndexedEntity.java]: Doesn't look related to this commit, does it?

File: engine/src/main/java/org/hibernate/search/bridge/builtin/ArrayBridge.java [hunk 3]
```diff
@@ -43,8 +45,9 @@ public FieldBridge getElementBridge() {
 	}
 
 	private void indexNotNullArray(String name, Object value, Document document, LuceneOptions luceneOptions) {
-		Object[] collection = (Object[]) value;
-		for ( Object entry : collection ) {
+		// Use Array.* to also support arrays of primitive values
+		for ( int i = 0, length = Array.getLength( value ) ; i < length ; ++i ) {
```
Comment (gsmet) [inline engine/src/main/java/org/hibernate/search/bridge/builtin/ArrayBridge.java]: I would vote for not using `Array` and introduce what's needed in `CollectionHelper` as we did for HV.

File: engine/src/main/java/org/hibernate/search/engine/spi/DocumentBuilderIndexedEntity.java (hunk 2 above)
Comment (yrodiere) [inline engine/src/main/java/org/hibernate/search/engine/spi/DocumentBuilderIndexedEntity.java]: It's related, since the commit makes it more likely that arrays will be used, but yeah, it addresses a different issue.

It's just nitpicking anyway: `initializeArray` is a no-op.

File: engine/src/main/java/org/hibernate/search/engine/spi/DocumentBuilderIndexedEntity.java (hunk 1 above)
Comment (yrodiere) [inline engine/src/main/java/org/hibernate/search/engine/spi/DocumentBuilderIndexedEntity.java]: Seriously?

**Comment by yrodiere** [comment]
@gsmet About using what's in [HV's CollectionHelper](https://github.com/hibernate/hibernate-validator/blob/master/engine/src/main/java/org/hibernate/validator/internal/util/CollectionHelper.java#L98), ok. I'll introduce another method to create an iterable, though, so we can use foreach loops.

**Comment by Sanne** [comment]
@Hibernate-CI retest this

**Comment by DavideD** [comment]
Jenkins, retest this, please

File: engine/src/main/java/org/hibernate/search/bridge/builtin/ArrayBridge.java (hunk 3 above)
Comment (yrodiere) [inline engine/src/main/java/org/hibernate/search/bridge/builtin/ArrayBridge.java]: Done.

**Comment by gsmet** [comment]
Rebased and merged, thanks!
